src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt

```diff
@@ -193,6 +208,55 @@ class JooqTouchstoneRepository(
         return records.map { mapCoverageSet(it) }
     }
 
+    override fun createCoverageSet(touchstoneVersionId: String,
+                                   vaccine: String,
+                                   activityType: ActivityType,
+                                   supportLevel: GAVISupportLevel): Int
+    {
+        try
+        {
+            val support = supportLevel.toString().toLowerCase()
+            val activity = activityType.toString().toLowerCase().replace('_', '-')
+            val record = this.dsl.newRecord(COVERAGE_SET).apply {
+                this.touchstone = touchstoneVersionId
+                this.name = "$vaccine: $vaccine, $support, $activity"
+                this.vaccine = vaccine
+                this.gaviSupportLevel = support
+                this.activityType = activity
+            }
+            record.store()
+            return record.id
+        }
+        catch(e: DataAccessException)
+        {
+            if (e.message?.contains("coverage_set_vaccine_fkey") == true) {
+                throw UnknownObjectError("vaccine", vaccine)
+            }
+            else throw e
+        }
+    }
+
+    override fun saveCoverageForTouchstone(touchstoneVersionId: String, records: List<CoverageRecord>)
+    {
+        dsl.batchStore(records).execute()
+    }
+
+    override fun newCoverageRowRecord(coverageSetId: Int,
+                                     country: String,
+                                     year: Int,
+                                     ageFrom: BigDecimal,
+                                     ageTo: BigDecimal,
+                                     target: BigDecimal?,
+                                     coverage: BigDecimal?) = this.dsl.newRecord(COVERAGE).apply {
+        this.coverageSet = coverageSetId
```

Comment:
gavi support at the coverage row level is redundant then?

Reply:
Thanks, you've drawn my attention to this which I had completely missed - I'm not actually sure from Xiang's scripts what the conditions are for gavi_support in the coverage table to be true. Following up with him to find out...